// @typescript-eslint/eslint-plugin/dist/rules/no-extra-parens.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const typescript_estree_1 = require("@typescript-eslint/typescript-estree");
const no_extra_parens_1 = __importDefault(require("eslint/lib/rules/no-extra-parens"));
const util = __importStar(require("../util"));
exports.default = util.createRule({
    name: 'no-extra-parens',
    meta: {
        type: 'layout',
        docs: {
            description: 'disallow unnecessary parentheses',
            category: 'Possible Errors',
            recommended: false,
        },
        fixable: 'code',
        schema: no_extra_parens_1.default.meta.schema,
        messages: no_extra_parens_1.default.meta.messages,
    },
    defaultOptions: ['all'],
    create(context) {
        const rules = no_extra_parens_1.default.create(context);
        return Object.assign({}, rules, {
            MemberExpression(node) {
                if (node.object.type !== typescript_estree_1.AST_NODE_TYPES.TSAsExpression) {
                    return rules.MemberExpression(node);
                }
            },
        });
    },
});
